Bugfix: Ensure Atomicity in Wallet Settings Updates from Chain Interface

ismaelsadeeq opened the pull request:
This PR fixes #30620.

As outlined in the issue, creating two wallets with `load_on_startup=true` simultaneously results in only one wallet being added to the startup file.

The current issue arises because the wallet settings update process involves:
1. Obtaining the settings value while acquiring the settings lock.
2. Modifying the settings value.
3. Overwriting the settings value while acquiring the settings lock again.

This sequence is not thread-safe. Different threads could modify the same base value simultaneously, overwriting data from other workers without realizing it.

The PR attempts to  fix this by modifying the chain interface's `updateRwSetting` method to accept a function that will be called with the settings reference. This function will either update or delete the setting and return an enum indicating whether the settings need to be overwritten in this or not.

Additionally, this PR introduces two new methods to the chain interface:
- `overwriteRwSetting`: This method replaces the setting with a new value.
Used in `VerifyWallets`
- `deleteRwSettings`: This method completely erases a specified setting. 
This method is currently used only in `overwriteRwSetting`.

These changes ensure that updates are race-free across all clients.

stickies-v commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -329,6 +329,39 @@ BOOST_FIXTURE_TEST_CASE(importwallet_rescan, TestChain100Setup)
+    WalletContext context;
+    context.chain = m_node.chain.get();
+    const int NUM_WALLETS = 5;
+    std::vector<std::thread> threads;
```
For stability, would ensure we start without any wallets (e.g. in case any defaults get added, will make it more visible why this test fails. Alternatively, could capture the count of wallets at the beginning of the test and use that to compare with later on, to make the test resilient to unrelated changes, but I think that might be a bit overkill.

```suggestion
    std::vector<std::thread> threads;
    // Since we're counting the number of wallets, ensure we start without any.
    BOOST_REQUIRE(context.chain->getRwSetting("wallet").isNull());
```

ismaelsadeeq replied:
done

stickies-v commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -329,6 +329,39 @@ BOOST_FIXTURE_TEST_CASE(importwallet_rescan, TestChain100Setup)
+    context.chain = m_node.chain.get();
+    const int NUM_WALLETS = 5;
+    std::vector<std::thread> threads;
+    {
```
nit: brief docstrings to at-a-glance help see how the test is structured:

<details>
<summary>git diff on 135eb1d5b1</summary>

```diff
diff --git a/src/wallet/test/wallet_tests.cpp b/src/wallet/test/wallet_tests.cpp
index ec9792f7a9..d434c1c207 100644
--- a/src/wallet/test/wallet_tests.cpp
+++ b/src/wallet/test/wallet_tests.cpp
@@ -337,6 +337,7 @@ BOOST_FIXTURE_TEST_CASE(write_wallet_settings_concurrently, TestingSetup)
     context.chain = m_node.chain.get();
     const int NUM_WALLETS = 5;
     std::vector<std::thread> threads;
+    // Add NUM_WALLETS wallets concurrently, ensure we end up with NUM_WALLETS.
     {
         for (int i{0}; i < NUM_WALLETS; i++) {
             threads.emplace_back([i, &context] {
@@ -349,6 +350,7 @@ BOOST_FIXTURE_TEST_CASE(write_wallet_settings_concurrently, TestingSetup)
         BOOST_CHECK_EQUAL(wallets.getValues().size(), NUM_WALLETS);
     }
     threads.clear();
+    // Remove NUM_WALLETS wallets concurrently, ensure we end up with 0 wallets.
     {
         for (int i{0}; i < NUM_WALLETS; i++) {
             threads.emplace_back([i, &context] {

```
</details>

ismaelsadeeq replied:
Added

stickies-v commented on src/wallet/wallet.h:
```diff
@@ -1112,10 +1112,10 @@ class WalletRescanReserver
 
 //! Add wallet name to persistent configuration so it will be loaded on startup.
-bool AddWalletSetting(interfaces::Chain& chain, const std::string& wallet_name);
+bool AddWalletSetting(WalletContext& context, const std::string& wallet_name);
```
Would be good to guard against lock contention (+ for `RemoveWalletSetting`):

<details>
<summary>git diff on 135eb1d5b1</summary>

```diff
diff --git a/src/wallet/wallet.h b/src/wallet/wallet.h
index ad2e2817c4..fd61b39c47 100644
--- a/src/wallet/wallet.h
+++ b/src/wallet/wallet.h
@@ -19,6 +19,7 @@
 #include <script/script.h>
 #include <support/allocators/secure.h>
 #include <sync.h>
+#include <threadsafety.h>
 #include <tinyformat.h>
 #include <uint256.h>
 #include <util/fs.h>
@@ -28,6 +29,7 @@
 #include <util/time.h>
 #include <util/ui_change_type.h>
 #include <wallet/crypter.h>
+#include <wallet/context.h>
 #include <wallet/db.h>
 #include <wallet/scriptpubkeyman.h>
 #include <wallet/transaction.h>
@@ -1112,10 +1114,10 @@ public:
 };
 
 //! Add wallet name to persistent configuration so it will be loaded on startup.
-bool AddWalletSetting(WalletContext& context, const std::string& wallet_name);
+bool AddWalletSetting(WalletContext& context, const std::string& wallet_name) EXCLUSIVE_LOCKS_REQUIRED(!context.wallets_mutex);
 
 //! Remove wallet name from persistent configuration so it will not be loaded on startup.
-bool RemoveWalletSetting(WalletContext& context, const std::string& wallet_name);
+bool RemoveWalletSetting(WalletContext& context, const std::string& wallet_name) EXCLUSIVE_LOCKS_REQUIRED(!context.wallets_mutex);
 
 struct MigrationResult {
     std::string wallet_name;

```
</details>

ismaelsadeeq replied:
Done.

stickies-v commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -329,6 +329,39 @@ BOOST_FIXTURE_TEST_CASE(importwallet_rescan, TestChain100Setup)
+                BOOST_CHECK_MESSAGE(RemoveWalletSetting(context, strprintf("wallet_%d", i)), strprintf("write wallet_%d failed", i));
+            });
+        }
+        for (auto& t : threads)
+            t.join();
```
style nit: should either use braces or multi-line, or move to one-line approach

```suggestion
        for (auto& t : threads) t.join();
```

ismaelsadeeq replied:
This is what was written initially. 
Clang format changed it to this format, which I thought was the recommended approach:
https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy

I've now switched to your suggestion.

ismaelsadeeq commented:
In draft now before addressing TSAN failure https://github.com/bitcoin/bitcoin/pull/30697/checks?check_run_id=29119417676

stickies-v commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -329,6 +329,39 @@ BOOST_FIXTURE_TEST_CASE(importwallet_rescan, TestChain100Setup)
+    WalletContext context;
+    context.chain = m_node.chain.get();
+    const int NUM_WALLETS = 5;
+    std::vector<std::thread> threads;
```
```suggestion
    std::vector<std::thread> threads;
    threads.reserve(NUM_WALLETS);
```

ismaelsadeeq replied:
Done.

ismaelsadeeq commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -329,6 +329,39 @@ BOOST_FIXTURE_TEST_CASE(importwallet_rescan, TestChain100Setup)
+    {
+        for (int i{0}; i < NUM_WALLETS; i++) {
+            threads.emplace_back([i, &context] {
+                BOOST_CHECK_MESSAGE(AddWalletSetting(context, strprintf("wallet_%d", i)), strprintf("write wallet_%d failed", i));
```
I suppose I should update to check for the boolean returned value of true not a string message.

 I will update to just `Assert` the returned value is true.

 I will test your suggestion, thanks

stickies-v replied:
As per https://github.com/bitcoin/bitcoin/pull/30697#discussion_r1727410617, I'm not sure capturing the offending thread adds value, since it won't be deterministic anyway? And we're already testing that all wallets were added/removed successfully? Also no objection to adding an `Assert`, though.

ismaelsadeeq replied:
Done thanks.

stickies-v commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -329,6 +329,39 @@ BOOST_FIXTURE_TEST_CASE(importwallet_rescan, TestChain100Setup)
+    {
+        for (int i{0}; i < NUM_WALLETS; i++) {
+            threads.emplace_back([i, &context] {
+                BOOST_CHECK_MESSAGE(AddWalletSetting(context, strprintf("wallet_%d", i)), strprintf("write wallet_%d failed", i));
```
I think TSAN is [failing](https://github.com/bitcoin/bitcoin/pull/30697/checks?check_run_id=29119417676) because of these `BOOST_CHECK_MESSAGE` statements. Since you're checking the count a bit further down already anyway, I think this is fine to remove. Exactly which thread failed doesn't really add information anyway, I think?

<details>
<summary>git diff on 135eb1d5b1</summary>

```diff
diff --git a/src/wallet/test/wallet_tests.cpp b/src/wallet/test/wallet_tests.cpp
index ec9792f7a9..651c6efdf2 100644
--- a/src/wallet/test/wallet_tests.cpp
+++ b/src/wallet/test/wallet_tests.cpp
@@ -340,7 +340,7 @@ BOOST_FIXTURE_TEST_CASE(write_wallet_settings_concurrently, TestingSetup)
     {
         for (int i{0}; i < NUM_WALLETS; i++) {
             threads.emplace_back([i, &context] {
-                BOOST_CHECK_MESSAGE(AddWalletSetting(context, strprintf("wallet_%d", i)), strprintf("write wallet_%d failed", i));
+                AddWalletSetting(context, strprintf("wallet_%d", i));
             });
         }
         for (auto& t : threads)
@@ -352,7 +352,7 @@ BOOST_FIXTURE_TEST_CASE(write_wallet_settings_concurrently, TestingSetup)
     {
         for (int i{0}; i < NUM_WALLETS; i++) {
             threads.emplace_back([i, &context] {
-                BOOST_CHECK_MESSAGE(RemoveWalletSetting(context, strprintf("wallet_%d", i)), strprintf("write wallet_%d failed", i));
+                RemoveWalletSetting(context, strprintf("wallet_%d", i));
             });
         }
         for (auto& t : threads)

```
</details>

ismaelsadeeq replied:
This is fixed.

stickies-v reviewed (commented):
Approach ACK

ismaelsadeeq commented:
Thanks @stickies-v for review, I've accepted all your suggestion and updated the branch.
I will force push after verifying TSan C.I passes locally

stickies-v commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -329,6 +330,44 @@ BOOST_FIXTURE_TEST_CASE(importwallet_rescan, TestChain100Setup)
+    {
+        for (int i{0}; i < NUM_WALLETS; i++) {
+            threads.emplace_back([i, &context] {
+                Assert(RemoveWalletSetting(context, strprintf("wallet_%d", i)));
```
nit: requires `#include <util/check.h>`

ismaelsadeeq replied:
Added

stickies-v commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -7,6 +7,7 @@
 #include <future>
 #include <memory>
 #include <stdint.h>
+#include <thread>
```
nit: while touching adjacent line, could help a neighbour out
```suggestion
#include <cstdint>
#include <future>
#include <memory>
#include <thread>
```

ismaelsadeeq replied:
Done.

stickies-v commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -329,6 +330,44 @@ BOOST_FIXTURE_TEST_CASE(importwallet_rescan, TestChain100Setup)
 }
 
+// This test verifies that wallet settings can be added and removed
+// Concurrently, ensuring no race conditions occur during either process.
```
```suggestion
// concurrently, ensuring no race conditions occur during either process.
```

ismaelsadeeq replied:
Taken

stickies-v reviewed (approved):
ACK 2cba0bed5be8a25e1635b7cde0a98285f64b5469, left non-blocking nits but happy to quickly re-ACK.

Would suggest updating the commit message to describe the change a bit better, add `bugfix` to the title, and remove the double `[]` and `:` notation:

> wallet: bugfix: lock wallet context before adding/removing wallet settings
> 
> When performing certain wallet operations simultaneously, a race condition can be
> triggered. For example, simultaneously loading two wallets can lead to only one
> wallet being added to the startup settings. Fix this by adding locks, and add test
> coverage.

stickies-v commented:
re-ACK 245dfa20bd37f78783adbd03897b7f144d71a7b8

furszy reviewed (commented):
I don't think using the context's wallet mutex is the best approach for this issue. The problem lies in how we update settings, which is not atomic. Today, the race is in the wallets flow, but tomorrow it could easily affect another settings option.

Updates are executed in three separate steps:
1) Obtain settings value while acquiring the settings lock.
2) Modify settings value.
3) Overwrite settings value while acquiring the settings lock.

This process is not thread-safe. Different threads could modify the same base value simultaneously, overwriting data from other workers without realizing it.

Given this context, I've implemented a different solution that makes the underlying settings update operation atomic: https://github.com/furszy/bitcoin-core/commit/de05ba1144bd6332d1d903d07e49463268412dc0. This commit also shortens the test code.

ismaelsadeeq reviewed (commented):
> Today, the race is in the wallets flow, but tomorrow it could easily affect another settings option.

I agree with this point. All clients that update settings will be victim to the same bug.

> Given this context, I've implemented a different solution that makes the underlying settings update operation atomic: [furszy@de05ba1](https://github.com/furszy/bitcoin-core/commit/de05ba1144bd6332d1d903d07e49463268412dc0). This commit also shortens the test code.

This seems like a great idea and an improvement to this PR. Clients will no longer need to lock a mutex before updating settings. I will take a look at  [furszy@de05ba1](https://github.com/furszy/bitcoin-core/commit/de05ba1144bd6332d1d903d07e49463268412dc0).
Thank you for your feedback @furszy

furszy commented:
Take this one: https://github.com/furszy/bitcoin-core/commit/b017949d83c1c47e6aea708ac89f0d81cead1064.
Have removed the `OptionModel` changes because those require further changes to the `Node` interface and implementation.

ismaelsadeeq committed:
wallet: bugfix: ensure atomicity in settings updates

- Settings updates were not thread-safe, as they were executed in
  three separate steps:

  1) Obtain settings value while acquiring the settings lock.
  2) Modify settings value.
  3) Overwrite settings value while acquiring the settings lock.

  This approach allowed concurrent threads to modify the same base value
  simultaneously, leading to data loss. When this occurred, the final
  settings state would only reflect the changes from the last thread
  that completed the operation, overwriting updates from other threads.

  Fix this by making the settings update operation atomic.

- Add test coverage for this behavior.

Co-authored-by: furszy <matiasfurszyfer@protonmail.com>

ismaelsadeeq reviewed (commented):
This force-push [diff](https://github.com/bitcoin/bitcoin/compare/245dfa20bd37f78783adbd03897b7f144d71a7b8..528d4e67e68350f451180a2356a635f67f02287f) is a complete change in approach from the previous one, which was simpler and more straightforward  245dfa20bd37f78783adbd03897b7f144d71a7b8 (i.e., locking the mutex before starting the settings update sequence). The commit was reviewed and acked by @stickies-v thanks

The new approach was written by @furszy which modifies the chain interface’s `updateRwSetting` method to receive an update function that is executed while the settings lock is held. The update function will returns an enum indicating whether the update was successful and if we should overwrite the settings.

I also think this approach is better and more sustainable, allowing other clients to avoid implementing their own thread control mechanisms for updating or deleting settings from the chain interface.

The changes in this diff include:
- Adding a `SettingsAction` enum, which indicates whether we want to write to the settings after the update.
- Introducing a `SettingsUpdate` type alias for the update function pointer.
- Updating `updateRwSetting` to receive a `SettingsUpdate` function pointer.
  - `overwriteRwSetting`: Replaces the setting with a new value.
  - `deleteRwSettings`: Completely erases a specified setting. This method is currently used only in `overwriteRwSetting`.
- Updating `AddWalletSetting` and `RemoveWalletSetting` to define and pass the update function to `updateRwSetting`.

The tests were also refactored to be a bit dry.

@furszy, I made some changes to your commit, including adjusting variable initializations to use auto-deduced types, adding type aliases for the function pointers, and making variable names more descriptive.

achow101 commented:
ACK 1b41d45d462d856a9d0b44ae0039bbb2cd78407c

furszy reviewed (commented):
self-code-ACK https://github.com/bitcoin/bitcoin/commit/1b41d45d462d856a9d0b44ae0039bbb2cd78407c

luke-jr commented:
Any reason not to just lock the mutex across the three steps?

ismaelsadeeq commented on src/interfaces/chain.h:
```diff
@@ -96,6 +96,17 @@ struct BlockInfo {
+    SKIP_WRITE
+};
+
+using SettingsUpdate = std::function<std::optional<interfaces::SettingsAction>(common::SettingsValue&)>;
```
I added it just to make the `updateRwSetting` method easier to read. I will update when their is need to retouch.

ismaelsadeeq commented:
> Any reason not to just lock the mutex across the three steps?

Yes the reasons where highlighted in https://github.com/bitcoin/bitcoin/pull/30697#pullrequestreview-2258765600 and https://github.com/bitcoin/bitcoin/pull/30697#pullrequestreview-2260433636 comments

stickies-v commented:
I wish this hadn't been merged so quickly as I was still working on reviewing this, but I understand getting this in before `28.x` branch off is helpful. I'm not a big fan of extending the `Chain` interface with non-chainstate related functions, that's a wrong direction imo. I think rebasing this PR on something like https://github.com/stickies-v/bitcoin/commit/53acb88072ce4450d4cc3f16c40b05d8382e7287 would have been a more elegant base.

Also note that this PR did not update `Node::updateRwSetting`, which now behaves more like `Chain::overwriteRwSetting`, and is theoretically vulnerable to the same race conditions as the ones addressed in this PR. Since this is only used in the GUI, triggering a race condition there seems less feasible though.

ryanofsky commented:
> I'm not a big fan of extending the `Chain` interface with non-chainstate related functions, that's a wrong direction imo.

It would be be nice to split up the Chain interface or just rename it to reflect way it is used, but it probably makes sense for wallet to use some interface class to read and write settings instead of using ArgsManager directly as in https://github.com/stickies-v/bitcoin/commit/53acb88072ce4450d4cc3f16c40b05d8382e7287. Wallet code originally did used ArgsManager directly until #22217, which changed it to use an interface class so wallet and node code could run in different processes and both could change settings without getting out of sync or corrupting the file. The wallet could be changed to use an independent settings file instead, and this would probably be a nice long term result. But defining interface classes for wallet code to go through instead of calling node functions and accessing node state directly has been an easy way to keep wallet code separate from node code without having to change wallet behavior.

furszy commented:
> Also note that this PR did not update Node::updateRwSetting, which now behaves more like Chain::overwriteRwSetting, and is theoretically vulnerable to the same race conditions as the ones addressed in this PR. Since this is currently only used by the GUI, triggering a race condition there seems less feasible though.

Yeah. I didn't touch it to avoid expanding this PR beyond what was necessary to fix the race. The GUI settings change events are executed sequentially on the main thread, so they aren't susceptible to this issue - at least not at this point in time -. We should definitely unify both `updateRwSetting` in a follow-up.

stickies-v commented on src/node/interfaces.cpp:
```diff
@@ -814,14 +814,32 @@ class ChainImpl : public Chain
+    bool deleteRwSettings(const std::string& name, bool write) override
+    {
+        args().LockSettings([&](common::Settings& settings) {
+            settings.rw_settings.erase(name);
```
Note: behaviour already existed prior to this PR, but this line (and thus also `overwriteRwSetting`) can throw if `name` is not an existing key, and would benefit from being documented in both function's docstring I think.

ryanofsky replied:
> Note: behaviour already existed prior to this PR, but this line (and thus also `overwriteRwSetting`) can throw if `name` is not an existing key, and would benefit from being documented in both function's docstring I think.

map::erase shouldn't actually throw anything, API was always intended to erase the setting if it exists otherwise do nothing

ismaelsadeeq replied:
I've tested this locally by passing an unknown settings name this does not throw.

stickies-v commented on src/node/interfaces.cpp:
```diff
@@ -814,14 +814,32 @@ class ChainImpl : public Chain
-            } else {
-                settings.rw_settings[name] = value;
-            }
+            auto* ptr_value = common::FindKey(settings.rw_settings, name);
+            // Create value if it doesn't exist
+            auto& value = ptr_value ? *ptr_value : settings.rw_settings[name];
```
I think this should be simplified to:

<details>
<summary>git diff on 1b41d45d46</summary>

```diff
diff --git a/src/node/interfaces.cpp b/src/node/interfaces.cpp
index 54b986c926..b1699bc7ae 100644
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -819,10 +819,8 @@ public:
     {
         std::optional<interfaces::SettingsAction> action;
         args().LockSettings([&](common::Settings& settings) {
-            auto* ptr_value = common::FindKey(settings.rw_settings, name);
-            // Create value if it doesn't exist
-            auto& value = ptr_value ? *ptr_value : settings.rw_settings[name];
-            action = update_settings_func(value);
+            auto& existing_value{settings.rw_settings[name]};
+            action = update_settings_func(existing_value);
         });
         if (!action) return false;
         // Now dump value to disk if requested

```
</details>

ryanofsky replied:
Agree this would be an improvement. But I think it would be better to change behavior and avoid writing null settings (for reasons in other comment). Would probably write as:

```c++
if (auto* value = common::FindKey(settings.rw_settings, name)) {
    action = update_settings_func(*value);
    if (value->isNull()) settings.rw_settings.erase(name);
} else {
    UniValue value;
    action = update_settings_func(*value);
    if (!value.isNull()) settings.rw_settings[name] = std::move(value);
}
```

ismaelsadeeq replied:
Taken in #30828

stickies-v commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -329,6 +329,40 @@ BOOST_FIXTURE_TEST_CASE(importwallet_rescan, TestChain100Setup)
+// concurrently, ensuring no race conditions occur during either process.
+BOOST_FIXTURE_TEST_CASE(write_wallet_settings_concurrently, TestingSetup)
+{
+    WalletContext context;
```
nit: there's no need for a `WalletContext` here:

<details>
<summary>git diff on 1b41d45d46</summary>

```diff
diff --git a/src/wallet/test/wallet_tests.cpp b/src/wallet/test/wallet_tests.cpp
index 12d5a3b3eb..7597765b24 100644
--- a/src/wallet/test/wallet_tests.cpp
+++ b/src/wallet/test/wallet_tests.cpp
@@ -333,12 +333,11 @@ BOOST_FIXTURE_TEST_CASE(importwallet_rescan, TestChain100Setup)
 // concurrently, ensuring no race conditions occur during either process.
 BOOST_FIXTURE_TEST_CASE(write_wallet_settings_concurrently, TestingSetup)
 {
-    WalletContext context;
-    context.chain = m_node.chain.get();
+    auto chain{m_node.chain.get()};
     const auto NUM_WALLETS{5};
 
     // Since we're counting the number of wallets, ensure we start without any.
-    BOOST_REQUIRE(context.chain->getRwSetting("wallet").isNull());
+    BOOST_REQUIRE(chain->getRwSetting("wallet").isNull());
 
     const auto& check_concurrent_wallet = [&](const auto& settings_function, int num_expected_wallets) {
         std::vector<std::thread> threads;
@@ -346,19 +345,19 @@ BOOST_FIXTURE_TEST_CASE(write_wallet_settings_concurrently, TestingSetup)
         for (auto i{0}; i < NUM_WALLETS; ++i) threads.emplace_back(settings_function, i);
         for (auto& t : threads) t.join();
 
-        auto wallets = context.chain->getRwSetting("wallet");
+        auto wallets = chain->getRwSetting("wallet");
         BOOST_CHECK_EQUAL(wallets.getValues().size(), num_expected_wallets);
     };
 
     // Add NUM_WALLETS wallets concurrently, ensure we end up with NUM_WALLETS stored.
-    check_concurrent_wallet([&context](int i) {
-        Assert(AddWalletSetting(*context.chain, strprintf("wallet_%d", i)));
+    check_concurrent_wallet([chain](int i) {
+        Assert(AddWalletSetting(*chain, strprintf("wallet_%d", i)));
     },
                             /*num_expected_wallets=*/NUM_WALLETS);
 
     // Remove NUM_WALLETS wallets concurrently, ensure we end up with 0 wallets.
-    check_concurrent_wallet([&context](int i) {
-        Assert(RemoveWalletSetting(*context.chain, strprintf("wallet_%d", i)));
+    check_concurrent_wallet([chain](int i) {
+        Assert(RemoveWalletSetting(*chain, strprintf("wallet_%d", i)));
     },
                             /*num_expected_wallets=*/0);
 }

```
</details>

ismaelsadeeq replied:
Removed in #30828

stickies-v reviewed (commented):
post-merge ACK 1b41d45d462d856a9d0b44ae0039bbb2cd78407c

I think some follow-up improvements should be made as per my review comments, but this PR addresses the race condition well.

> Wallet code originally did used ArgsManager directly until https://github.com/bitcoin/bitcoin/pull/22217, which changed it to use an interface class so wallet and node code could run in different processes and both could change settings without getting out of sync or corrupting the file.

Thanks for the context, I was wondering why things are the way they are at the moment. It makes sense in a multi-process context, but yeah, I don't think these functions should be in `Chain`, but it may currently be the least bad option. The developer notes also state we shouldn't be implementing any new functionality in interface methods, which we're violating here:

> - Interface method definitions should wrap existing functionality instead of
  implementing new functionality. Any substantial new node or wallet
  functionality should be implemented in [`src/node/`](../src/node/) or
  [`src/wallet/`](../src/wallet/) and just exposed in
  [`src/interfaces/`](../src/interfaces/) instead of being implemented there,
  so it can be more modular and accessible to unit tests.

> The wallet could be changed to use an independent settings file instead, and this would probably be a nice long term result.

This was my initial thought too, but too big of a change for a bugfix PR indeed.

stickies-v commented:
> Wallet code originally did used ArgsManager directly until https://github.com/bitcoin/bitcoin/pull/22217, which changed it to use an interface class so wallet and node code could run in different processes and both could change settings without getting out of sync or corrupting the file.

Actually, doesn't the fact that the Rw methods on both `Node` and `Chain` interfaces directly use `ArgsManager` violate that principle too?

ryanofsky commented:
> Actually, doesn't the fact that the Rw methods on both Node and Chain interfaces directly use ArgsManager violate that principle too? 

This is ok because of the way multiprocess code works. When node and wallet code are running in the same process, and the wallet calls a virtual `interfaces::Chain` method, this just invokes the `node::ChainImpl` implementation of that method in `src/node/interfaces.cpp`. But when node and wallet code are running in different processes, and wallet code calls an `interfaces::Chain` method, it calls an IPC implementation of that method that forwards the method call and arguments to the node process, and then invokes the `node::ChainImpl` implemention of the method in the node process. So all settings reads and writes happen through the node process ArgsManager object and there is not more than one process accessing the settings file.

> The developer notes also state we shouldn't be implementing any new functionality in interface methods, which we're violating here

I think this is a borderline case. Ideally wrapper methods in src/node/interfaces.cpp and src/wallet/interfaces.cpp are just a few lines long and consist of glue code that doesn't implement any real functionality. I think that is basically the case with the methods in this PR, though though they do implement some options and features, and maybe those could be moved from ChainImpl to ArgsManager so the ArgsManager interface is more complete and glue code is simpler.

stickies-v commented:
> This is ok because of the way multiprocess code works ... there is not more than one process accessing the settings file.

Thank you, that explanation actually clears up a lot, appreciate the thoughtful response, as always.

> and maybe those could be moved from ChainImpl to ArgsManager so the ArgsManager interface is more complete and glue code is simpler.

and it would also allow code re-use across both interfaces.

ryanofsky commented:
Thanks, I think a natural cleanup or followup to this PR could move the `interfaces::Node` settings methods (`isSettingIgnored`, `getPersistentSetting`, `updateRwSetting`, `forceSetting`, and `resetSettings`) and the `interfaces::Chain` settings methods (`getSetting`, `getSettingList`, `getRwSetting`, `updateRwSetting`, `overwriteRwSetting`, `deleteRwSettings`) into a new `interfaces::Settings` class and add a new `interfaces::Init` method returning it.

In the longer run, I think wallet processes should not need to access the settings interface directly, either because they store their own JSON settings detached from node settings, or because the setting that tracks which wallets to load at startup is managed by gui or node processes instead of wallet processes (i.e. the list of wallets to load would be passed to the `WalletLoaderImpl::load()` method, instead of it reading settings and deciding which wallets to load itself).

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -814,14 +814,32 @@ class ChainImpl : public Chain
+    {
+        if (value.isNull()) return deleteRwSettings(name, write);
+        return updateRwSetting(name, [&](common::SettingsValue& settings) {
+            settings = std::move(value);
```
IMO it's not great that this new API allows setting null values when previous API did not. Allowing null values to be set could encourage setting null values, which could encourage using null values, and treating them differently than missing values, which is probably not a good outcome. (Or even if it is good, is something that should be thought about more clearly, not just introduced in a bugfix.)

ismaelsadeeq replied:
Agreed, I see no reason for allowing a null value.
Removed in #30828

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -814,14 +814,32 @@ class ChainImpl : public Chain
+    {
+        if (value.isNull()) return deleteRwSettings(name, write);
+        return updateRwSetting(name, [&](common::SettingsValue& settings) {
+            settings = std::move(value);
```
Agree rvalue reference would be an improvement, less for the reason that it would make the function safer to use, and more for the reason that it would make the function more convenient to use, because lvalue references cannot bind to temporary or literal values.

But even better than an rvalue reference would just be a plain value like `common::SettingsValue value` because this would allow the caller to decide whether they want to move or copy, instead of forcing the caller to always move.

ismaelsadeeq replied:
Taken your suggestion @ryanofsky this is fixed in #30828

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -814,14 +814,32 @@ class ChainImpl : public Chain
+        // Now dump value to disk if requested
+        return *action == interfaces::SettingsAction::SKIP_WRITE || args().WriteSettingsFile();
+    }
+    bool overwriteRwSetting(const std::string& name, common::SettingsValue& value, bool write) override
```
Seems like as long as `SettingsAction` enum exists this function and deleteRwSettings below should probably take `SettingsAction action` parameters instead of `bool write` parameters to be more consistent and explicit.

ismaelsadeeq replied:
Done in #30828

ryanofsky reviewed (commented):
Just leaving some notes in case there are more changes to settings api later.

ismaelsadeeq commented:
Thanks for your reviews, @stickies-v and @ryanofsky.  
I've addressed the review comments in #30828.


> Thanks. I think a natural cleanup or follow-up to this PR could involve moving the `interfaces::Node` settings methods (`isSettingIgnored`, `getPersistentSetting`, `updateRwSetting`, `forceSetting`, and `resetSettings`) and the `interfaces::Chain` settings methods (`getSetting`, `getSettingList`, `getRwSetting`, `updateRwSetting`, `overwriteRwSetting`, `deleteRwSettings`) into a new `interfaces::Settings` class and adding a new `interfaces::Init` method to return it.

I have started working on this in [this commit](https://github.com/bitcoin/bitcoin/commit/5de29069690c68f3fc17a654617bc8b6f567e128). 
I believe this should come after #30828, but let me know if you'd prefer combining the two.
